sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+        // seems to actually verify this anywhere?
+
+        Ok(())
+    }
+
+    pub async fn get_account(
+        &self,
+        address: AccountAddress,
+    ) -> Result<Response<Option<AccountView>>> {
+        Ok(self
+            .send_via_batch(MethodRequest::GetAccount((address,)))
+            .await?
+            .and_then(MethodResponse::try_into_get_account)?)
+    }
+
+    /// Send a single request via `VerifyingClient::batch`.
+    async fn send_via_batch(&self, request: MethodRequest) -> Result<Response<MethodResponse>> {
+        let mut responses = self.batch(vec![request]).await?.into_iter();
+        let response = match responses.next() {
+            Some(response) => response,
+            None => bail!("response batch is empty"),
+        };
+        ensure!(responses.next().is_none(), "too many responses");
+        response
+    }
+
+    pub async fn batch(
+        &self,
+        requests: Vec<MethodRequest>,
+    ) -> Result<Vec<Result<Response<MethodResponse>>>> {
+        // transform each request into verifying sub-request batches
+        let batch = VerifyingBatch::from_batch(requests);
+
+        // flatten and collect sub-request batches into flat list of requests
+        let mut requests = batch.collect_requests();
+        // append get_state_proof request
+        let start_version = self.version();
+        requests.push(MethodRequest::get_state_proof(start_version));
+
+        // actually send the batch
+        let responses = self.inner.batch(requests).await?;
+
+        // TODO(philiphayes): REMOVE. for now, just convert the errors until we
+        // use the real error type.
+        let mut responses = responses
+            .into_iter()
+            .map(|result| result.map_err(anyhow::Error::new))
+            .collect::<Vec<_>>();
+
+        // extract and verify the state proof
+        let (state_proof_response, state) = responses.pop().unwrap()?.into_parts();
+        let state_proof_view = state_proof_response.try_into_get_state_proof()?;
+        let state_proof: StateProof = (&state_proof_view).try_into()?;
+
+        // check the response metadata matches the state proof
+        let latest_li = state_proof.0.ledger_info();
+        ensure!(state.version == latest_li.version());
+        ensure!(state.timestamp_usecs == latest_li.timestamp_usecs());
+
+        // try to ratchet our trusted state using the state proof
+        self.verify_and_ratchet(&state_proof)?;
+
+        // remote says we're too far behind and need to sync. we have to throw
+        // out the batch since we can't verify any proofs
+        if state_proof.1.more {
+            bail!("needs sync");
+        }
+
+        // unflatten subresponses, verify, and collect into one response per request
+        batch.validate_responses(start_version, &state, &state_proof, responses)
+    }
+}
+
+struct VerifyingBatch {
+    requests: Vec<VerifyingRequest>,
```

Comment:
it took me a while to realize this is not `Vec<MethodRequest>` and the collect_requests added a lot subrequests.
I feel it may not worth the effort to make such complex structure for batching multiple original requests. 
In practice, I only see 1 or 2 use cases requires batch those requests for non-verifying APIs.
Make it for one original request will simplify the implementation a lot, and make it easier to maintain.

Reply:
I agree the implementation is a bit complicated. I tried to make the key implementation point self-contained (see `verified_get_account`) so implementing a new API call via verifying APIs (that also support batching) shouldn't feel too different from the same implementation that doesn't support batching.

 I'm not sure how else to support request batching though (at least without adding an extra RTT and potential load balancer issues, as discussed in the doc) : (

Are you suggesting we don't support request batches for verifying clients?

Reply:
As a first step maybe we don't support batching, just to simplify the design (since its already a complex thing to get right)? Just talking out loud, I don't have a strong opinion either way

Reply:
If we didn't have to deal with request batching, we would remove the `VerifyingBatch` + `VerifyingRequest` and instead `get_account` would probably look something like:

```rust
pub async fn get_account(&self, address: AccountAddress) -> Result<Option<AccountView>> {
    let responses = self.inner.batch(vec![
        MethodRequest::GetStateProof(self.version()),
        MethodRequest::GetAccountStateWithProof(diem_root_address(), None, None),
        MethodRequest::GetAccountStateWithProof(address, None, None),
    ]).await?;

    // verify all responses have same version + timestamp + chain id and pull them out

    // collect inner responses

    match &responses {
        [MethodResponse::GetStateProof(state_proof), MethodResponse::GetAccountStateWithProof(diem_root), MethodResponse::GetAccountStateWithProof(account)] => {
            // verify state proof version + timestamp match response version + timestamp

            self.verify_and_ratchet(&state_proof)?;

            // everything in `verified_get_account` callback
        },
        _ => bail!(),
    }

}
```

or if we didn't want to repeat all that common verification stuff each time, it would probably look like `VerifyingRequest` and `batch` does now

Reply:
I think this is better starting point. Batch requests are not scalable as we have to limit batch size (current value is 20).

Reply:
Personally, I'm in favor of maintaining batching support:

1. If our API is going to support batching, then all of our clients _must_ support batching. If we're not going to support batching in some clients, then why support batching in the service at all? It's going to cause tons of confusion for users if some client interfaces implement arbitrary subsets of functionality.

2. I'm intending for this `VerifyingClient` interface to be mostly transparent; in other words, a user should be able to drop-in replace the non-verifying `Client` with `VerifyingClient`, which is no longer true if the `VerifyingClient` can't match the full `Client` interface. Similarly, if we ever abstract `Client` as a trait, the two main implementation points are `batch` and `send_one`. Without supporting batches, `VerifyingClient` can't impl the trait.

3. Batching amortizes `GetStateProof` request overhead.

4. Without batching, atomic queries of the chain state (across multiple requests) always require an extra RTT to first get the most recent version and then pin the subsequent queries to that version.

As far as implementation complexity, the `VerifyingBatch` logic is a one-time cost. Adding support for each request type (`get_account`, `get_events`, etc...) doesn't require batch-aware logic and should look the same regardless of whether batching is supported.

Reply:
Ok.

Please write enough tests to cover all of cases.